src/flashcart/ed64/ed64_ll.c

```diff
@@ -0,0 +1,409 @@
+#include <malloc.h>
+#include <libdragon.h>
+#include "utils/utils.h"
+#include "ed64_ll.h"
+
+
+
+/* ED64 configuration registers base address  */
+#define ED64_CONFIG_REGS_BASE (0xA8040000)
+
+typedef enum {
+    // REG_CFG = 0,
+    // REG_STATUS = 1,
+    // REG_DMA_LENGTH = 2,
+    // REG_DMA_RAM_ADDR = 3,
+    // REG_MSG = 4,
+    // REG_DMA_CFG = 5,
+    // REG_SPI = 6,
+    // REG_SPI_CFG = 7,
+    // REG_KEY = 8,
+    REG_SAV_CFG = 9,
+    // REG_SEC = 10, /* Sectors?? */
+    // REG_FPGA_VERSION = 11, /* Altera (Intel) MAX */
+    // REG_GPIO = 12,
+
+} ed64_registers_t;
+
+void pi_initialize (void);
+void pi_initialize_sram (void);
+void pi_dma_from_cart (void* dest, void* src, unsigned long size);
+void pi_dma_to_cart (void* dest, void* src, unsigned long size);
+void pi_dma_from_sram (void *dest, unsigned long offset, unsigned long size);
+void pi_dma_to_sram (void* src, unsigned long offset, unsigned long size);
+void pi_dma_from_cart_safe (void *dest, void *src, unsigned long size);
+
+void ed64_ll_set_sdcard_timing (void);
+
+
+#define SAV_EEP_ON 1
+#define SAV_SRM_ON 2
+#define SAV_EEP_SIZE 4
+#define SAV_SRM_SIZE 8
+
+#define SAV_RAM_BANK 128
+#define SAV_RAM_BANK_APPLY 32768
+
+uint32_t ed64_ll_reg_read (uint32_t reg);
+void ed64_ll_reg_write (uint32_t reg, uint32_t data);
+
+uint8_t ed64_ll_sram_bank;
+ed64_save_type_t ed64_ll_save_type;
+
+
+uint32_t ed64_ll_reg_read (uint32_t reg) {
+
+    *(volatile uint32_t *) (ED64_CONFIG_REGS_BASE);
+    return *(volatile uint32_t *) (ED64_CONFIG_REGS_BASE + reg * 4);
+
+}
+
+void ed64_ll_reg_write (uint32_t reg, uint32_t data) {
+
+    *(volatile uint32_t *) (ED64_CONFIG_REGS_BASE);
```

Comment:
Why is this needed, this dummy read is probably optimized away by the compiler anyway